zproject/backends.py

```diff
@@ -1630,6 +1631,93 @@ def redirect_deactivated_user_to_login(realm: Realm, email: str) -> HttpResponse
+) -> int | None:
+    """
+    Syncs user attributes based on the SOCIAL_AUTH_SYNC_ATTRS_DICT setting.
+    Only supports:
+    1. Syncing the role. This is plumbed through to user creation, so can be
+       used to immediately create new users with their role set based on an attribute
+       provided by the IdP.
+    2. Syncing custom attributes. This isn't supported for user creation,
+       so they'll only be synced during the user's next login, not during
+       signup.
+    """
+    # This is only supported for SAML right now, though the design
+    # is meant to be easy to extend this to other backends if desired.
+    # Unlike with LDAP, here we can only do syncing during the authentication
+    # flow, as that's when the data is provided and we don't have a way to query
+    # for it otherwise.
+    assert backend.name == "saml"
+
+    attrs_by_backend = settings.SOCIAL_AUTH_SYNC_ATTRS_DICT.get(realm.subdomain, {})
+    profile_field_name_to_attr_name = attrs_by_backend.get(backend.name, {})
+    if not extra_attrs or not profile_field_name_to_attr_name:
+        return None
+
+    user_id = None
+    if user_profile is not None:
+        user_id = user_profile.id
+
+    custom_profile_field_name_to_value = {}
+    new_role = None
+    for field_name, attr_name in profile_field_name_to_attr_name.items():
+        if field_name == "role":
+            ROLE_NAME_TO_TYPE = {
+                "owner": UserProfile.ROLE_REALM_OWNER,
+                "administrator": UserProfile.ROLE_REALM_ADMINISTRATOR,
+                "moderator": UserProfile.ROLE_MODERATOR,
+                "member": UserProfile.ROLE_MEMBER,
+                "guest": UserProfile.ROLE_GUEST,
+            }
+            ROLE_TYPE_TO_NAME = {v: k for k, v in ROLE_NAME_TO_TYPE.items()}
+            attr_value = extra_attrs.get(attr_name)
+            if not attr_value:
+                continue
+            if attr_value not in ROLE_NAME_TO_TYPE:
+                backend.logger.warning(
+                    "Ignoring unsupported role value %s for user %s in SOCIAL_AUTH_SYNC_ATTRS_DICT",
+                    attr_value,
+                    user_id,
+                )
+                continue
+            new_role = ROLE_NAME_TO_TYPE[attr_value]
+        elif field_name.startswith("custom__"):
+            custom_profile_field_name_to_value[field_name.removeprefix("custom__")] = (
+                extra_attrs.get(attr_name)
+            )
+        else:
+            backend.logger.warning(
+                "Ignoring unsupported UserProfile field %s in SOCIAL_AUTH_SYNC_ATTRS_DICT",
+                field_name,
+            )
+
+    if user_profile is None:
+        # We don't support user creation with custom profile fields, so just
+        # return role so that it can be plumbed through to the signup flow.
+        if new_role is not None:
+            backend.logger.info("Returning role %s for user creation", ROLE_TYPE_TO_NAME[new_role])
+        return new_role
```

Comment:
I guess it might not be a hard follow-up to plumb through a custom profile field mapping as well, just to be able to delete the "only syncs on second login" weirdness. Might be reasonable to do while we're at it.

Reply:
Hmm, this isn't super obvious, because we don't currently have a good way to plumb the custom profile field values information through to user creation. We could add that to `PreregistrationUser`, maybe in the form of a simple map `dict[str, <Possible values>]`? `"{<custom profile field name>: <intended value>"}` in a new field `PreregistrationUser.custom_profile_field_values_map`?

Or maybe that should be keyed by proper `ForeignKey`s to `CustomProfileField` objects?

Once we start plumbing this information through on `PreregistrationUser` objects, we then have to be aware of some race conditions like e.g. the intended `CustomProfileField` getting deleted (or modified) after the info was saved in the object, but before the user registered. Especially since a `PreregistrationUser` can live for quite a while before actually ending up being used if the user delays finishing the signup - so this isn't an unlikely race.

So not sure, should we just proceed with either of these approaches, or worth a discussion in #backend?

Reply:
Probably worth discussing in #backend. Might not be worth doing; fundamentally SCIM is the system intended for this problem statement with SAML.